controllers/vspherevm_controller.go

```diff
@@ -269,6 +271,25 @@ func (r vmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.R
 		if err := vmContext.Patch(ctx); err != nil {
 			reterr = kerrors.NewAggregate([]error{reterr, err})
 		}
+
+		// Wait until VSphereVM is updated in the cache if the `.Status.TaskRef` field changes.
+		// Note: We have to do this because otherwise using a cached client in current state could
+		// return a stale state of a VSphereVM we just patched (because the cache might be stale).
+		if vmContext.VSphereVM.Status.TaskRef != originalTaskRef {
+			err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+				key := ctrlclient.ObjectKey{Namespace: vmContext.VSphereVM.GetNamespace(), Name: vmContext.VSphereVM.GetName()}
+				cachedMD := &infrav1.VSphereVM{}
+				if err := r.Client.Get(ctx, key, cachedMD); err != nil {
+					return false, err
+				}
+				return originalTaskRef != cachedMD.Status.TaskRef, nil
+			})
+			if reterr != nil {
+				reterr = kerrors.NewAggregate([]error{reterr, err})
+			} else {
+				reterr = err
+			}
```

Comment:
```suggestion
			reterr = kerrors.NewAggregate([]error{reterr, err})
```
Aggregate will take care of it